pkg/model/stack_serializer.go

```diff
@@ -0,0 +1,839 @@
+// Copyright 2020 The Okteto Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/okteto/okteto/pkg/log"
+	apiv1 "k8s.io/api/core/v1"
+)
+
+//Stack represents an okteto stack
+type StackRaw struct {
+	Version    string                 `yaml:"version,omitempty"`
+	Name       string                 `yaml:"name"`
+	Xname      string                 `yaml:"x-name,omitempty"`
+	Namespace  string                 `yaml:"namespace,omitempty"`
+	Xnamespace string                 `yaml:"x-namespace,omitempty"`
+	Services   map[string]*ServiceRaw `yaml:"services,omitempty"`
+	Endpoints  map[string][]Endpoint  `yaml:"endpoints,omitempty"`
+
+	// Docker-compose not implemented
+	Networks *WarningType `yaml:"networks,omitempty"`
+	Volumes  *WarningType `yaml:"volumes,omitempty"`
+	Configs  *WarningType `yaml:"configs,omitempty"`
+	Secrets  *WarningType `yaml:"secrets,omitempty"`
+
+	Warnings []string
+}
+
+//Service represents an okteto stack service
+type ServiceRaw struct {
+	Deploy          *DeployInfoRaw     `yaml:"deploy,omitempty"`
+	Build           *BuildInfo         `yaml:"build,omitempty"`
+	CapAdd          []apiv1.Capability `yaml:"cap_add,omitempty"`
+	CapDrop         []apiv1.Capability `yaml:"cap_drop,omitempty"`
+	Command         Args               `yaml:"command,omitempty"`
+	Entrypoint      Command            `yaml:"entrypoint,omitempty"`
+	EnvFiles        []string           `yaml:"env_file,omitempty"`
+	Environment     *RawMessage        `yaml:"environment,omitempty"`
+	Expose          *RawMessage        `yaml:"expose,omitempty"`
+	Image           string             `yaml:"image,omitempty"`
+	Labels          *RawMessage        `json:"labels,omitempty" yaml:"labels,omitempty"`
+	Annotations     map[string]string  `json:"annotations,omitempty" yaml:"annotations,omitempty"`
+	Xannotations    map[string]string  `json:"x-annotations,omitempty" yaml:"x-annotations,omitempty"`
+	Ports           []Port             `yaml:"ports,omitempty"`
```

Comment:
I would have two different structs, one for `PortRaw` and one for `Port`. `PortRaw` has containerPort and hostPort. And as part of the traslating the port to the stack struct, we just get the containerPort, and set `service.Public` if a hostPort is not nil